ambry-store/src/main/java/com/github/ambry/store/PersistentIndex.java

```diff
@@ -478,12 +479,15 @@ void changeIndexSegments(List<File> segmentFilesToAdd, Set<Offset> segmentsToRem
   }
 
   /**
-   * Adds a new entry to the index
+   * Adds a new entry to the index. For recovery, only the index segments belongs to last log segment should be put into
+   * journal, other wise it won't be compacted and we will facing the issue when last log segment is empty but journal size
+   * is non-zero.
    * @param entry The entry to be added to the index
    * @param fileSpan The file span that this entry represents in the log
+   * @param isForRecovery Determine if the addToIndex method is for recovery.
    * @throws StoreException
    */
-  void addToIndex(IndexEntry entry, FileSpan fileSpan) throws StoreException {
+  void addToIndex(IndexEntry entry, FileSpan fileSpan, boolean isForRecovery) throws StoreException {
```

Comment:
all of the function invocations has the last parameter set to false except for one. I suggest in this case, we can add a new method with `isForRecovery` being false by default. In this way, we only have to change one function invocation.

Reply:
BTW, can you change `isForRecovery` to `recovering`, just to use the same variable name as BlobStoreCompactor.commit.